### PR TITLE
Fix PHP 8 unit test GHA job; remove PHPUnit dependency and instead use PHPUnit polyfills package

### PR DESCRIPTION
**Pull request by pierlon**
## Summary

Note: Each commit in this PR represents an individual set of changes and so it's recommended to go through each commit instead of viewing all the files that changed.

This PR fixes the PHP 8 unit test GHA, along with some other changes that will allow us to easily adopt newer versions of PHPUnit when the time comes.

### Changes

- Replace deprecated PHPUnit assertions with their successors. This will allow us to easily switch to newer versions of PHPUnit in the future without hassle. (https://github.com/ampproject/amp-wp/pull/6525/commits/d8468568f58089c5e25102545bda6fbce6f1e191, https://github.com/ampproject/amp-wp/pull/6525/commits/d1438ad2482e0a284827736f3479d7947df46441, https://github.com/ampproject/amp-wp/pull/6525/commits/c82893208edd05103be726ce2b7a28c8d1377392, https://github.com/ampproject/amp-wp/pull/6525/commits/bf34961733aa7fa70ecb23edd1a43d1cedcd37bd, https://github.com/ampproject/amp-wp/pull/6525/commits/cce7f32a20dd687a0239ff822125fcd6d8d17868)

- Remove `AssertContainsCompatibility` trait as it is now unnecessary (https://github.com/ampproject/amp-wp/pull/6525/commits/0ec6eaf1cd41139f0368ff804434f976ff280397)

- Remove the `phpunit/phpunit` Composer package and use [`yoast/wp-test-utils`](https://github.com/Yoast/wp-test-utils) to manage it instead. This allows us to remove the hard PHPUnit 5.7 requirement and delegate that responsibility to `wp-test-utils` which will install the most recent compatible version of PHPUnit depending on the active PHP version (https://github.com/ampproject/amp-wp/pull/6525/commits/47a1e6bef3ee41db8f95e3bebb705bdf9347dfe2)

- Remove no longer necessary `composer.json` patch used to run PHPUnit tests on PHP 8 (https://github.com/ampproject/amp-wp/pull/6525/commits/25ae209eff9698a8546d4a6956649b6f835990bc)

- Update all tests to use the `Yoast\WPTestUtils\WPIntegration\TestCase` instead of `WP_UnitTestCase`. This will still allow us to use all the WP Core test utilities as it extends `WP_UnitTestCase` and will load the correct PHPUnit polyfills when necessary (https://github.com/ampproject/amp-wp/pull/6525/commits/ebfd6e95889889d5762df2dff500715cd1f24df4)

- Patch the `phpunit/phpunit-mock-objects` package to prevent PHP warnings from occurring on PHP 8, which without it would cause some tests to fail (https://github.com/ampproject/amp-wp/pull/6525/commits/ffdcc781768f01241066ab19d4bf7f925141903d)

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: tests/php/src/PairedUrlTest.php [hunk 1]
```diff
@@ -16,7 +16,6 @@ public function setUp() {
 		$this->instance = $this->injector->make( PairedUrl::class );
 	}
 
-	/** @covers ::__construct() */
```
Comment (westonruter) [inline tests/php/src/PairedUrlTest.php:19]: Does this just not do anything?

File: tests/php/test-amp-style-sanitizer.php [hunk 2]
```diff
@@ -2037,7 +2036,7 @@ public function get_external_stylesheet_data() {
 				],
 				'expected_styles' => [],
 				'expected_errors' => [ AMP_Style_Sanitizer::STYLESHEET_FETCH_ERROR ],
-				'cached_data'    => new CachedResponse(
+				'expected_cached_response'    => new CachedResponse(
```
Comment (westonruter) [inline tests/php/test-amp-style-sanitizer.php:2039]: Does `TestCase` map associative arrays to named positional args?

File: tests/php/test-includes-admin-functions.php [hunk 3]
```diff
@@ -125,7 +126,7 @@ public function get_data_for_test_amp_get_customizer_url() {
 					Option::THEME_SUPPORT => AMP_Theme_Support::READER_MODE_SLUG,
 					Option::READER_THEME  => ReaderThemes::DEFAULT_READER_THEME,
 				],
-				function () {
+				'assert' => function () {
```
Comment (westonruter) [inline tests/php/test-includes-admin-functions.php]: Was this causing an error before or is it just a bug? It's strange that there's an associative array mixed with a numerical index.

**Review by westonruter — APPROVED** [review]
Amazing. 😍

File: tests/php/src/PairedUrlTest.php (hunk 1 above)
Comment (pierlon) [reply tests/php/src/PairedUrlTest.php:19]: When running PHPUnit with coverage it complained that `__construct()` didn't exist in the class, so in short yes.

File: tests/php/test-amp-style-sanitizer.php (hunk 2 above)
Comment (pierlon) [reply tests/php/test-amp-style-sanitizer.php:2039]: In PHP 8 it does.

File: tests/php/test-includes-admin-functions.php (hunk 3 above)
Comment (pierlon) [reply tests/php/test-includes-admin-functions.php]: In PHP 8 it maps to a named parameter.

**Comment by schlessera** [comment]
I'd recommend using a plugin-specific `TestCase` class, instead of coupling every single test file against an external dependency.

See https://github.com/ampproject/amp-toolbox-php/blob/main/tests/src/TestCase.php for a simple example.

This decouples the tests from whatever compatibility layer is needed.

**Pull request merged**